### PR TITLE
Lecture 1 setup and product dao

### PR DESCRIPTION
**Pull request by nestifff**

### CODE REVIEW

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 1]
```diff
@@ -1,25 +1,94 @@
...
+        save(new Product("nokia3310", "Nokia 3310", new BigDecimal(70), usd, 100, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Nokia/Nokia%203310.jpg"));
+        save(new Product("palmp", "Palm Pixi", new BigDecimal(170), usd, 30, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Palm/Palm%20Pixi.jpg"));
+        save(new Product("simc56", "Siemens C56", new BigDecimal(70), usd, 20, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20C56.jpg"));
+        save(new Product("simc61", "Siemens C61", new BigDecimal(80), usd, 30, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20C61.jpg"));
+        save(new Product("simsxg75", "Siemens SXG75", new BigDecimal(150), usd, 40, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20SXG75.jpg"));
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: Please make sure the images are displayed.

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 2]
```diff
@@ -1,25 +1,94 @@
...
+            if (existedProductWithThisId == null) {
+                addProduct(product);
+
+            } else {
+                existedProductWithThisId.setCode(product.getCode());
+                existedProductWithThisId.setCurrency(product.getCurrency());
+                existedProductWithThisId.setDescription(product.getDescription());
+                existedProductWithThisId.setImageUrl(product.getImageUrl());
+                existedProductWithThisId.setPrice(product.getPrice());
+                existedProductWithThisId.setStock(product.getStock());
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: Can we make code a bit more readable by moving this to a separate method like populateExitingProduct()?

File: src/main/java/com/es/phoneshop/web/ProductListPageServlet.java [hunk 3]
```diff
@@ -1,7 +1,10 @@
 package com.es.phoneshop.web;
 
+import com.es.phoneshop.model.product.ArrayListProductDao;
 import com.es.phoneshop.model.product.Product;
+import com.es.phoneshop.model.product.ProductDao;
 
+import javax.servlet.ServletConfig;
 import javax.servlet.ServletException;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/web/ProductListPageServlet.java]: Remove unused imports.

File: src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java [hunk 4]
```diff
@@ -15,7 +19,27 @@ public void setup() {
...
+        assertTrue(testProduct.getId() > 0);
+        assertNotNull(productDao.getProduct(testProduct.getId()));
+        assertSame(productDao.getProduct(testProduct.getId()), testProduct);
+    }
+
+    @Test(expected = NoSuchElementException.class)
+    public void testDeleteProduct() {
+        Long id = 1L;
+        productDao.delete(id);
+        productDao.getProduct(id);
```
Comment (SergeiStarikov) [inline src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java]: Please add more tests for each aspect of functionality: get, find, save, delete product.
Add more testing for corner cases: what if we search for null product? What is we try to save the product without ID? Delete product that does not exist? etc.

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 5]
```diff
@@ -1,25 +1,94 @@
...
+
+    public ArrayListProductDao() {
+        this.products = new ArrayList<>();
+        saveSampleProducts();
+    }
+
     @Override
-    public Product getProduct(Long id) {
-        throw new RuntimeException("Not implemented");
+    public synchronized Product getProduct(Long id) throws NoSuchElementException {
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: Can you think about using other synchronization mechanism instead of synchronized keyword?

**Review by SergeiStarikov — APPROVED** [review]

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java (hunk 1 above)
Comment (nestifff) [reply src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: done

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java (hunk 2 above)
Comment (nestifff) [reply src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: done

File: src/main/java/com/es/phoneshop/web/ProductListPageServlet.java (hunk 3 above)
Comment (nestifff) [reply src/main/java/com/es/phoneshop/web/ProductListPageServlet.java]: done

File: src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java (hunk 4 above)
Comment (nestifff) [reply src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java]: done

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java (hunk 5 above)
Comment (nestifff) [reply src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: done

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 6]
```diff
@@ -1,25 +1,163 @@
...
+                    addProduct(product);
+                    return true;
+                } else {
+                    changeExistingProduct(existedProductWithThisId, product);
+                    return false;
+                }
+
+            } finally {
+                lock.writeLock().unlock();
+            }
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: Can we do the following?

```
public boolean save(Product product) {
    **lock.writeLock().lock();**
    try {
        if (product.getId() == null) {
            **// No explicit locking here**
            addProduct(...)
        } else {
            **// No explicit locking here**
            getProduct(id)...
            changeExistingProduct(...)
            addProduct(...)
        }
    } catch(...) {
        ...
    } finally {
        **lock.writeLock().unlock();**
    }

}
```

This way we use the following approach:
lock writeLock
addProduct
unlock writeLock

getProduct() method contains read locking so it's not needed to lock the readLock in save method.

**Review by SergeiStarikov — APPROVED** [review]

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java (hunk 6 above)
Comment (nestifff) [reply src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: done
